### DISCUSSION TITLE
Can't understand Inputs for k-NN interpolation

### DISCUSSION THREAD

**Discussion by atale480** [pyg-team/pytorch_geometric#6860] [Q&A] [answered]
I am trying to implement PointNet++. I'm not able to understand what does parameters for knn_interpolate from documentation.
First parameter X is easy but remaining 2 I didn't understand. Can anyone explain me in simple terms what are those parameters?
[Link for documentation of knn_interpolate](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.unpool.knn_interpolate.html#torch_geometric.nn.unpool.knn_interpolate)

**Comment by wsad1** [comment, marked as answer]
`knn_interpolate` usually operates on point clouds. point clouds have a feature vector `x` (the first argument) and a position vector `pos_x` and `pos_y`. The position vectors are used to calculate distance between nodes. This distance is later used as a weight how important a node is while interpolating node features.

**Reply by atale480** [reply]
What does pos_x and pos_y mean that I'm not able to understand

**Reply by wsad1** [reply]
`pos_x` is the position of the nodes in some `d` dimensional space.
`pos_y` is the upsampled node position matrix. 
See this [example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/pointnet2_segmentation.py) code, that uses `knn_interpolate` for segmentation.

**Reply by atale480** [reply]
Thanks for answer!!! I have just one query.
As I can see in the PointNet++ paper architecture that during segmentation author regenerated original shape using interpolation. To get this point only and not go till segmentation do I need to stop at [MLP layer](https://github.com/pyg-team/pytorch_geometric/blob/ad2fe06b5f5a69ddcf201e3c21257c9abc92c2e1/examples/pointnet2_segmentation.py#L60), or do I need Linear layer also as done in the [example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/pointnet2_segmentation.py) you mentioned.

**Reply by wsad1** [reply]
In that example the `linear` layers aren't used, need to remove them from init.
So you can stop at the mlp layer.
